subprojects/plugins/src/main/java/org/gradle/api/plugins/internal/JvmPluginsHelper.java

```diff
@@ -123,12 +124,20 @@ public static void configureForSourceSet(final SourceSet sourceSet, final Source
     private static void configureForSourceSet(final SourceSet sourceSet, final SourceDirectorySet sourceDirectorySet, AbstractCompile compile, final Project target) {
         compile.setDescription("Compiles the " + sourceDirectorySet.getDisplayName() + ".");
         compile.setSource(sourceSet.getJava());
+        boolean useDefaultCompilationOrder = target.getConvention().getPlugin(JavaPluginConvention.class).getDefaultCompilationOrder();
```

Comment:
So my understanding is that this works only because we use `tasks.register` to land here. Is there any way we can protect against someone calling this in an eager fashion?

Reply:
See #11513